### PR TITLE
HPCC-22960 Check mismatched file type reads

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday please review.

NB: I have not changed NewDiskRead, that may need revisiting for similar checks when it is generalized to handle csv, xml etc.

**Comment by ghalliday** [comment]
Was this triggered by a particular problem that was hit?  I'm wondering how this change would be generalized when we have pluggable file formats.

**Comment by jakesmith** [comment]
I user hit it, incorrectly trying to read a CSV as a FLAT file.
They targeted Thor, which would have given a mismatch error, but because it was a simple test and they weren't using pickBestEngine=false, it ran on HThor and failed with a deserialization error, which didn't mean much to them.

**Comment by jakesmith** [comment]
>I'm wondering how this change would be generalized when we have pluggable file formats.

In general there should be a match.
But we have a couple of exception at the moment:

1) Reading any file format as a CSV
2) Reading any file format if FLAT and fixed width.
NB: it won't try to translate by default unless file type is also flat.

Perhaps new pluggable file formats might lend themselves to read arbitrary data too, in same way as CSV can be applied. So perhaps whether strict (must match read type) or not should be an option of the pluggable file reader interface.

**Comment by jakesmith** [comment]
There was a couple of issues exposed by the regression.
One was that there the warning reported to the workunit wasn't consistent, both in terms of error number and severity level (should have been warning). One of the tests that was failing had a commented out #onwarning, probably because it looked like it was necessary, but the warning was being reported as info only and not in the output.

I've made these warnings consistent now.
I've opened a separate JIRA re. consolidating error code numbers in 1 place for all engines for common errors like this: https://track.hpccsystems.com/browse/HPCC-22976

**Comment by jakesmith** [comment]
@ghalliday - please re-review (perhaps after regression suite passes)

File: thorlcr/thorutil/thormisc.cpp [hunk 1]
```diff
@@ -805,7 +805,15 @@ void reportExceptionToWorkunit(IConstWorkUnit &workunit,IException *e, ErrorSeve
...
     }
-} 
+}
+
+void reportExceptionToWorkunitCheckIgnore(IConstWorkUnit &workunit, IException *e, ErrorSeverity severity)
+{
+    ErrorSeverity mappedSeverity = workunit.getWarningSeverity(e->errorCode(), SeverityWarning);
+    if (SeverityIgnore == mappedSeverity)
+        return;
+    reportExceptionToWorkunit(workunit, e, severity);
```
Comment (ghalliday) [inline thorlcr/thorutil/thormisc.cpp]: Should this pass mappedSeverity instead of severity?

**Comment by ghalliday** [comment]
@jakesmith one minor question

File: thorlcr/thorutil/thormisc.cpp (hunk 1 above)
Comment (jakesmith) [reply thorlcr/thorutil/thormisc.cpp]: agree, have changed, and squashed change in.

**Comment by jakesmith** [comment]
@ghalliday - see addressed comment.

**Pull request merged**